### PR TITLE
More listings refactoring

### PR DESCRIPTION
**Pull request by citizen428**
## What type of PR is this? (check all applicable)

- [x] Refactor

## Description

This is the 3 PR in the series of listing code refactorings I started a while ago. I just noticed it's been sitting in draft for a while, not quite sure how that happened, I apparently just forgot about it 🤦 

## Added/updated tests?

- [X] No, because no observable behavior changed.

## [Forem core team only] How will this change be communicated?

- [X] This change does not need to be communicated, and this is why not: refactoring only

### CODE REVIEW

File: app/controllers/api/v0/listings_controller.rb [hunk 1]
```diff
@@ -10,7 +10,6 @@ class ListingsController < ApiController
       before_action :authenticate_with_api_key_or_current_user!, only: %i[create update]
       before_action :authenticate_with_api_key_or_current_user, only: %i[show]
       before_action :set_cache_control_headers, only: %i[index show]
-      before_action :set_listing, only: %i[update]
```
Comment (citizen428) [inline app/controllers/api/v0/listings_controller.rb:13]: This is no longer needed here as the concern now uses an `included` hook to set up this action.

File: app/controllers/concerns/listings_toolkit.rb [hunk 2]
```diff
@@ -4,14 +4,19 @@ module ListingsToolkit
 
   MANDATORY_FIELDS_FOR_UPDATE = %i[body_markdown title tag_list].freeze
 
-  def set_listing
+  included do
+    before_action :set_and_authorize_listing, only: %i[update]
```
Comment (citizen428) [inline app/controllers/concerns/listings_toolkit.rb:8]: At least for `:update` we can set up this action from the concern.

File: app/controllers/concerns/listings_toolkit.rb [hunk 3]
```diff
@@ -4,14 +4,19 @@ module ListingsToolkit
...
+  end
+
+  def set_and_authorize_listing
```
Comment (citizen428) [inline app/controllers/concerns/listings_toolkit.rb:11]: Updated this to include authorization. Every time `set_listing` was used it was implicitly in a `before_action` the first line of the method was `authorize @listing`.

File: app/controllers/concerns/listings_toolkit.rb [hunk 4]
```diff
@@ -4,14 +4,19 @@ module ListingsToolkit
...
     @listing = Listing.find(params[:id])
+    authorize @listing
   end
 
   def rate_limit?
     begin
       rate_limit!(:listing_creation)
-    rescue StandardError => e
+    rescue ::RateLimitChecker::LimitReached => e
```
Comment (citizen428) [inline app/controllers/concerns/listings_toolkit.rb:19]: I felt like we should check for the correct error here. The method is narrowly scoped so we shouldn't just swallow unrelated errors.

File: app/controllers/concerns/listings_toolkit.rb [hunk 5]
```diff
@@ -124,7 +121,7 @@ def first_publish(cost)
     author = @listing.author
     available_user_credits = author.is_a?(Organization) ? current_user.credits.unspent.size : 0
 
-    if author.credits.unspent.size >= cost
+    if author.enough_credits?(cost)
```
Comment (citizen428) [inline app/controllers/concerns/listings_toolkit.rb:124]: Reusing the existing `User#enough_credits?` method. I'd prefer eventually completely removing this from `User`.

File: app/controllers/concerns/listings_toolkit.rb [hunk 6]
```diff
@@ -120,11 +98,28 @@ def update
     process_after_update
   end
 
+  private
+
+  def set_and_authorize_listing
```
Comment (citizen428) [inline app/controllers/concerns/listings_toolkit.rb:103]: Every action that uses `set_listing` calls `authorize @listing` as its first line so I added that to the action.

File: app/controllers/concerns/listings_toolkit.rb [hunk 7]
```diff
@@ -120,11 +98,28 @@ def update
...
+    @listing = Listing.find(params[:id])
+    authorize @listing
+  end
+
+  def rate_limit?
+    begin
+      rate_limit!(:listing_creation)
+    rescue ::RateLimitChecker::LimitReached => e
```
Comment (citizen428) [inline app/controllers/concerns/listings_toolkit.rb:111]: Given the name and scope of this method catching a more specific error seemed like a good choice.

File: app/services/listings/bump.rb [hunk 8]
```diff
@@ -8,29 +8,20 @@ def initialize(listing, user:)
...
       @cost = listing.cost
       @user = user
-      @org = Organization.find_by(id: listing.organization_id)
     end
 
     def call
-      purchaser = [@org, @user].detect { |who| who&.enough_credits?(@cost) }
-      return false unless purchaser
+      @listing.purchase(@user) do |purchaser|
+        ActiveRecord::Base.transaction do
```
Comment (jeremyf) [inline app/services/listings/bump.rb:15]: Definitely like that this is in a transaction.  Is "enough_credits" the right response variable?  (Non-blocking, I was expecting a variable named "bought")

**Review by jeremyf — APPROVED** [review]

File: app/services/listings/bump.rb (hunk 8 above)
Comment (citizen428) [reply app/services/listings/bump.rb:15]: If the `@listings.bump` call further down the transaction fails (it won't) they wouldn't be bought. But I see what you're saying and may rename it.

File: app/controllers/listings_controller.rb [hunk 9]
```diff
@@ -52,7 +52,6 @@ def index
     @listings_json = @listings.to_json(INDEX_JSON_OPTIONS)
     @displayed_listing_json = @displayed_listing.to_json(INDEX_JSON_OPTIONS)
 
-    # TODO: [mkohl] Can we change this to listings-#{params[:category]}?
```
Comment (citizen428) [inline app/controllers/listings_controller.rb:55]: Finally removed this comment because it really doesn't matter.

**Review by jeremyf — APPROVED** [review]

**Review by joshpuetz — APPROVED** [review]

**Pull request merged**